File: scripts/foil/rebuildCardsIndex.py
#!/usr/bin/env python3
"""Rebuild data/<pack>/cards-index.json (v1) after layout migration."""

from __future__ import annotations

import json
import re
import sys
from datetime import datetime, timezone
from pathlib import Path

REPO = Path(__file__).resolve().parents[2]
BUNDLE_RE = re.compile(r"^([a-z0-9.-]+)_([a-z]{2,4})_(\d{3})$", re.I)
PRINT_RE = re.compile(r"^lorcana:([a-z0-9.]+)-(.+)$", re.I)


def rebuild_lorcana() -> None:
    cards_dir = REPO / "data/lorcana/cards"
    index = {
        "version": 1,
        "pack": "lorcana",
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "cards": {},
    }
    if not cards_dir.is_dir():
        print("no lorcana cards")
        return
    for set_dir in sorted(cards_dir.iterdir()):
        if not set_dir.is_dir():
            continue
        set_code = set_dir.name
        for lang_dir in sorted(set_dir.iterdir()):
            if not lang_dir.is_dir():
                continue
            lang = lang_dir.name
            for card_dir in sorted(lang_dir.iterdir()):
                if not card_dir.is_dir():
                    continue
                card = card_dir.name
                lang_files: dict[str, str] = {}
                for f in card_dir.iterdir():
                    if not f.is_file():
                        continue
                    stem = f.stem.lower()
                    name = f.name
                    if stem == "art":
                        lang_files["art"] = name
                    elif stem == "thumb":
                        lang_files["thumb"] = name
                    elif stem == "mask" or stem.startswith("mask"):
                        lang_files["mask"] = name
                    elif "second" in stem and "varnish" in stem:
                        lang_files["secondVarnishMask"] = name
                    elif "varnish" in stem:
                        lang_files["varnishMask"] = name
                    elif stem == "back":
                        lang_files["back"] = name
                parts = card.split("-", 1)
                if len(parts) == 2 and parts[1]:
                    print_key = f"lorcana:{set_code}-{parts[0]}-{parts[1]}"
                else:
                    print_key = f"lorcana:{set_code}-{card}"
                entry = index["cards"].setdefault(
                    print_key, {"set": set_code, "card": card, "langs": {}}
                )
                entry["langs"][lang] = lang_files
    out = REPO / "data/lorcana/cards-index.json"
    body = json.dumps(index, separators=(",", ":")) + "\n"
    out.write_text(body, encoding="utf-8")
    print(f"lorcana: {len(index['cards'])} prints → {out}")


def rebuild_pokemon() -> None:
    cards_dir = REPO / "data/pokemon/cards"
    index = {
        "version": 1,
        "pack": "pokemon",
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "cards": {},
    }
    if not cards_dir.is_dir():
        print("no pokemon cards")
        return
    n = 0
    for set_dir in cards_dir.iterdir():
        if not set_dir.is_dir():
            continue
        set_code = set_dir.name
        for lang_dir in set_dir.iterdir():
            if not lang_dir.is_dir():
                continue
            lang = lang_dir.name
            for card_dir in lang_dir.iterdir():
                if not card_dir.is_dir():
                    continue
                card = card_dir.name
                stem = f"{set_code}_{lang}_{card}"
                if not BUNDLE_RE.match(stem):
                    continue
                lang_files: dict = {}
                variants: dict = {}
                for f in card_dir.iterdir():
                    if not f.is_file():
                        continue
                    low = f.name.lower()
                    if low.startswith("art."):
                        lang_files["art"] = f.name
                    elif low in ("mask.webp", "mask.png", "mask.jpg"):
                        lang_files["mask"] = f.name
                    elif low.startswith("mask-ph."):
                        variants.setdefault("ph", {})["mask"] = f.name
                    elif low.startswith("mask-mph."):
                        variants.setdefault("mph", {})["mask"] = f.name
                    elif low.startswith("mask-sph."):
                        variants.setdefault("sph", {})["mask"] = f.name
                    elif low.startswith("etch."):
                        lang_files["etch"] = f.name
                    elif low.startswith("back."):
                        lang_files["back"] = f.name
                if variants:
                    lang_files["variants"] = variants
                index["cards"][stem] = {
                    "set": set_code,
                    "card": card,
                    "langs": {lang: lang_files},
                }
                n += 1
                if n % 10000 == 0:
                    print(f"  … {n}")
    out = REPO / "data/pokemon/cards-index.json"
    out.write_text(json.dumps(index, separators=(",", ":")) + "\n", encoding="utf-8")
    print(f"pokemon: {len(index['cards'])} stems → {out}")


def main() -> None:
    rebuild_lorcana()
    rebuild_pokemon()


if __name__ == "__main__":
    main()

File: scripts/foil/rebuildCardsIndex.ts
#!/usr/bin/env npx tsx
/**
 * Rebuild `data/<pack>/cards-index.json` (v1) after layout migration.
 */
import fs from "node:fs";
import path from "node:path";

import { parsePrintKey } from "../src/core/identify/printKey";
import type { CardsIndexV1 } from "../src/effects/cardsIndex";
import {
  packCardsDir,
  packCardsIndexPath,
  repoRoot,
} from "../scripts/lib/foilPaths";

const BUNDLE_RE =
  /^([a-z0-9.-]+)_([a-z]{2,4})_(\d{3})(?:_[a-z]+)?$/i;

function rebuildLorcana(repo: string): void {
  const cardsDir = packCardsDir(repo, "lorcana");
  const index: CardsIndexV1 = {
    version: 1,
    pack: "lorcana",
    generatedAt: new Date().toISOString(),
    cards: {},
  };

  // Prefer catalog.sqlite export when available
  const dbPath = path.join(repo, "data/lorcana/catalog.sqlite");
  if (fs.existsSync(dbPath)) {
    // Dynamic require avoided — walk disk for filenames, keys from print dirs via parse
  }

  if (!fs.existsSync(cardsDir)) {
    console.warn("no lorcana cards dir");
    return;
  }

  for (const set of fs.readdirSync(cardsDir)) {
    const setDir = path.join(cardsDir, set);
    if (!fs.statSync(setDir).isDirectory() || set === "back.webp") continue;
    if (set.startsWith("back.")) continue;
    for (const lang of fs.readdirSync(setDir)) {
      const langDir = path.join(setDir, lang);
      if (!fs.statSync(langDir).isDirectory()) continue;
      for (const card of fs.readdirSync(langDir)) {
        const cardDir = path.join(langDir, card);
        if (!fs.statSync(cardDir).isDirectory()) continue;
        const files = fs.readdirSync(cardDir).filter((f) =>
          fs.statSync(path.join(cardDir, f)).isFile(),
        );
        const langFiles: Record<string, string> = {};
        for (const f of files) {
          const stem = f.replace(/\.[^.]+$/, "").toLowerCase();
          if (stem === "art") langFiles.art = f;
          else if (stem === "thumb") langFiles.thumb = f;
          else if (stem === "mask" || stem.startsWith("mask")) langFiles.mask = f;
          else if (stem.startsWith("varnish") && stem.includes("second")) {
            langFiles.secondVarnishMask = f;
          } else if (stem.startsWith("varnish")) langFiles.varnishMask = f;
          else if (stem === "back") langFiles.back = f;
        }
        // Rebuild printKey
        const parts = card.split("-");
        const number = parts[0]!;
        const grouping = parts.length > 1 ? parts.slice(1).join("-") : null;
        const printKey = grouping
          ? `lorcana:${set}-${number}-${grouping}`
          : `lorcana:${set}-${number}`;
        if (!parsePrintKey(printKey)) {
          console.warn("skip bad printKey", printKey);
          continue;
        }
        const entry = index.cards[printKey] ?? {
          set,
          card,
          langs: {},
        };
        entry.langs[lang] = langFiles;
        index.cards[printKey] = entry;
      }
    }
  }

  const out = packCardsIndexPath(repo, "lorcana");
  fs.writeFileSync(out, `${JSON.stringify(index)}\n`);
  console.log(`lorcana cards-index: ${Object.keys(index.cards).length} prints → ${out}`);
}

function rebuildPokemon(repo: string): void {
  const cardsDir = packCardsDir(repo, "pokemon");
  const index: CardsIndexV1 = {
    version: 1,
    pack: "pokemon",
    generatedAt: new Date().toISOString(),
    cards: {},
  };
  if (!fs.existsSync(cardsDir)) {
    console.warn("no pokemon cards dir yet");
    return;
  }
  for (const set of fs.readdirSync(cardsDir)) {
    const setDir = path.join(cardsDir, set);
    if (!fs.statSync(setDir).isDirectory()) continue;
    if (set.startsWith("back.")) continue;
    for (const lang of fs.readdirSync(setDir)) {
      const langDir = path.join(setDir, lang);
      if (!fs.statSync(langDir).isDirectory()) continue;
      for (const card of fs.readdirSync(langDir)) {
        const cardDir = path.join(langDir, card);
        if (!fs.statSync(cardDir).isDirectory()) continue;
        const stem = `${set}_${lang}_${card}`;
        if (!BUNDLE_RE.test(stem)) continue;
        const files = fs.readdirSync(cardDir).filter((f) =>
          fs.statSync(path.join(cardDir, f)).isFile(),
        );
        const langFiles: Record<string, string> = {};
        const variants: Record<string, { mask?: string; etch?: string }> = {};
        for (const f of files) {
          const low = f.toLowerCase();
          if (low.startsWith("art.")) langFiles.art = f;
          else if (low === "mask.webp" || low === "mask.png") langFiles.mask = f;
          else if (low.startsWith("mask-ph.")) {
            variants.ph = { ...(variants.ph ?? {}), mask: f };
          } else if (low.startsWith("mask-mph.")) {
            variants.mph = { ...(variants.mph ?? {}), mask: f };
          } else if (low.startsWith("mask-sph.")) {
            variants.sph = { ...(variants.sph ?? {}), mask: f };
          } else if (low.startsWith("etch.")) langFiles.etch = f;
          else if (low.startsWith("back.")) langFiles.back = f;
        }
        if (Object.keys(variants).length) langFiles.variants = variants as never;
        index.cards[stem] = {
          set,
          card,
          langs: { [lang]: langFiles },
        };
      }
    }
  }
  const out = packCardsIndexPath(repo, "pokemon");
  fs.writeFileSync(out, `${JSON.stringify(index)}\n`);
  console.log(`pokemon cards-index: ${Object.keys(index.cards).length} stems → ${out}`);
}

const repo = repoRoot();
rebuildLorcana(repo);
rebuildPokemon(repo);
